falcon/request.py

```diff
@@ -446,33 +111,152 @@ class Request:
         '_media_error',
         'is_websocket',
     )
-
-    _cookies = None
-    _cookies_collapsed = None
-    _cached_if_match = None
-    _cached_if_none_match = None
+    _cookies: Optional[Dict[str, List[str]]] = None
+    _cookies_collapsed: Optional[Dict[str, str]] = None
+    _cached_if_match: MissingOr[Optional[List[Union[ETag, Literal['*']]]]] = MISSING
+    _cached_if_none_match: MissingOr[Optional[List[Union[ETag, Literal['*']]]]] = (
+        MISSING
```

Comment:
Could we keep `None` for now? This PR isn't the smallest to follow, so ideally we should minimize the introduced runtime changes outside of typing. We probably need to profile things again otherwise (although these cached headers are probably not the most/important sensitive part; we should move the ones that are less frequently used to being cached in some `_state` attribute to increase perf even more).

Reply:
I don't think we need to profile things for this change, since a comparison using `is` has the same time.

missing is used when the final value can include none. in those cases having a default of None is suboptimal like mentioned in the previous TODO.

I can create a new PR for this, but seemed so small a change that made sense to do inline with the typing one

Reply:
I don't think this is true for cythonized code, but since these methods are not very critical for core performance of the framework, I agree it is no big deal.

Just for reference
```python
_OBJECT = object()


def func1(obj):
    return obj is None


def func2(obj):
    return obj is _OBJECT
```
:arrow_down: 
```
.....................
func1 (None is None): Mean +- std dev: 31.5 ns +- 0.4 ns
.....................
func1 (_OBJECT is None): Mean +- std dev: 32.0 ns +- 2.0 ns
.....................
func2 (None is _OBJECT): Mean +- std dev: 37.0 ns +- 0.6 ns
.....................
func2 (_OBJECT is _OBJECT): Mean +- std dev: 36.9 ns +- 1.0 ns
```

This is because `is None` in Cython can be implemented as checking a C pointer vs `NULL`, whereas a global or local constant still needs to be resolved from name first.

But as said, I don't think this difference is enough to bother. But I do want to underscore these constants when used in runtime, so `_Missing`, `_MISSING`, `_UNSET`, etc.

Reply:
as mentioned my idea was to worry about the public/private in a final cleanup PR. but if you prefer it here I can change it here.

personally I think all the properties could be re-written using a memorizing property since request has dict in any case

Reply:
OK, but let's not forget about that cleanup then!

Reply:
roger. Should we also try the momoizing prop approach?

I think it would end up faster both in init than in use

Reply:
Dunno, we could maybe focus on getting 4.0 out the door first, and then circle back on rebuilding benchmarks, Codspeed, and optimizations.

I was thinking to add a private `._state` attribute, `object` or `dict` (whatever suits better), and then store less used cached header properties there. Also store stuff like the state of dependent middleware which we would deprecate and provide a shim for.

Reply:
if we keep `__dict__` in request, memoized properties are all faster, since they avoid the property after the first run

Reply:
Re cleanup/privatizing, I added a note on #1457.

Reply:
but sure let's just keep it in mind